### PR TITLE
Calendar repeating - first round, Google Calendar Exporter

### PR DESCRIPTION
**Pull request by olsona**
After looking around at various options (ical4j, dmfs/lib-recur, etc) I determined that there aren't any easily available external libraries that do what we want: turn strings into RRules, RDates, and ExDates.  Since we're not trying to interpret the recurrence rules but instead merely store them, it's not unreasonable to write our own lightweight class to do just that.  

There is still a lot of testing and cleanup to do.  This PR is just a check to make sure I'm not going too far down a wrong path.

### CODE REVIEW

File: extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/calendar/GoogleCalendarExporter.java [hunk 1]
```diff
@@ -81,13 +85,39 @@ private static CalendarAttendeeModel transformToModelAttendee(EventAttendee atte
     return new CalendarEventModel.CalendarEventTime(offsetDateTime, dateTime.getDate() != null);
   }
 
+  private static RecurrenceRule getRecurrenceRule(List<String> ruleStrings) {
+    RecurrenceRule.Builder ruleBuilder = new RecurrenceRule.Builder();
+    for (String st : ruleStrings) {
+      String[] split = st.split(":");
+      String type = split[0];
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/calendar/GoogleCalendarExporter.java]: Let's add a precondition to ensure there are two parts or meets a certain format requirement

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+public class RecurrenceRule {
+
+  public static final String RRULE = "RRULE";
+  public static final String RDATE = "RDATE";
+  public static final String EXDATE = "EXDATE";
+
+  private RRule rRule;
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: can these all be final

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+   * @return an {@link RRule}
+   */
+  public static RRule parseRRuleString(String rRuleString) {
+    RRule.Builder builder = RRule.builder();
+
+    List<String> components = Arrays.asList(rRuleString.split(";"));
+    Map<ByRule, String> byRuleMapInput = new HashMap<>();
+    for (String property : components) {
+      String[] split = property.split("=");
+      String key = split[0];
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: Precondition that check we have 2 parts?

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+   * RDATE:)
+   * @return an {@link RDate}
+   */
+  public static RDate parseRDateString(String rDateString) {
+    RDate.Builder builder = RDate.builder();
+    StringTokenizer stringTokenizer = new StringTokenizer(rDateString, ":");
+    String token = stringTokenizer.nextToken();
+    while (stringTokenizer.hasMoreTokens()) {
+      String[] split = token.split("=");
+      String key = split[0];
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: same here

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    return builder.build();
+  }
+
+  public static ExDate parseExDateString(String exDateString) {
+    ExDate.Builder builder = ExDate.builder();
+    StringTokenizer stringTokenizer = new StringTokenizer(exDateString, ":");
+    String token = stringTokenizer.nextToken();
+    while (stringTokenizer.hasMoreTokens()) {
+      String[] split = token.split("=");
+      String key = split[0];
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: same here

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+      token = stringTokenizer.nextToken();
+    }
+    builder.rdtval(Arrays.asList(token.split(",")));
+
+    return builder.build();
+  }
+
+  public static ExDate parseExDateString(String exDateString) {
+    ExDate.Builder builder = ExDate.builder();
+    StringTokenizer stringTokenizer = new StringTokenizer(exDateString, ":");
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: Do we have to check if the string is null or empty here or before this is called?

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+      public abstract Builder until(String until);
+
+      public abstract Builder count(Integer count);
+
+      public abstract Builder interval(Integer interval);
+
+      public abstract Builder byRuleMap(Map<ByRule, String> byRuleMap);
+
+      public abstract Builder wkst(Day wkst);
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: What does this stand for? Let's rename or add comment.

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+      public abstract RRule build();
+    }
+  }
+
+  @AutoValue
+  public abstract static class RDate {
+
+    @Nullable abstract TimeType value();
+
+    @Nullable abstract String tzidparam();
```
Comment (holachuy) [inline portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: let's add comment here

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java (hunk 2 above)
Comment (olsona) [reply portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: Done.

File: portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java (hunk 7 above)
Comment (olsona) [reply portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/calendar/RecurrenceRule.java]: Added comment at the top about this.

**Pull request merged**